rhel/vex/parser.go

```diff
@@ -0,0 +1,450 @@
+				v := &claircore.Vulnerability{
+					Updater:            u.Name(),
+					Name:               name,
+					Description:        desc,
+					Issued:             v.ReleaseDate,
+					Links:              strings.Join(links, " "),
+					Severity:           "Unknown",
+					NormalizedSeverity: claircore.Unknown,
+				}
+				return v
+			}
+			// We're only bothered about known_affected and fixed,
+			// not_affected and under_investigation are ignored.
+			fixedVulns, err := creator.fixedVulnerabilities(ctx, v, protoVuln)
+			if err != nil {
+				return nil, nil, err
+			}
+			out[name] = fixedVulns
+			knownAffectedVulns, err := creator.knownAffectedVulnerabilities(ctx, v, protoVuln)
+			if err != nil {
+				return nil, nil, err
+			}
+			out[name] = append(out[name], knownAffectedVulns...)
+		}
+	}
+	vulns := []*claircore.Vulnerability{}
+	for _, vs := range out {
+		vulns = append(vulns, vs...)
+	}
+
+	return vulns, deleted, nil
+}
+
+type repoCache struct {
+	cache map[string]*claircore.Repository
+}
+
+func NewRepoCache() *repoCache {
+	return &repoCache{
+		cache: make(map[string]*claircore.Repository),
+	}
+}
+
+func (rc *repoCache) Get(cpe cpe.WFN) *claircore.Repository {
+	if r, ok := rc.cache[cpe.String()]; ok {
+		return r
+	}
+	r := &claircore.Repository{
+		CPE:  cpe,
+		Name: cpe.String(),
+		Key:  repoKey,
+	}
+	rc.cache[cpe.String()] = r
+	return r
+}
+
+type productCache struct {
+	cache map[string]*csaf.Product
+}
+
+func NewProductCache() *productCache {
+	return &productCache{
+		cache: make(map[string]*csaf.Product),
+	}
+}
+
+func (pc *productCache) Get(c *csaf.CSAF, productID string) *csaf.Product {
+	if p, ok := pc.cache[productID]; ok {
+		return p
+	}
+	p := c.ProductTree.FindProductByID(productID)
+	pc.cache[productID] = p
+	return p
+
+}
+
+func NewCreator(vulnName, vulnLink string, c *csaf.CSAF, pc *productCache, rc *repoCache) *creator {
+	return &creator{
+		vulnName:       vulnName,
+		vulnLink:       vulnLink,
+		uniqueVulnsIdx: make(map[string]int),
+		c:              c,
+		pc:             pc,
+		rc:             rc,
+	}
+}
+
+type creator struct {
+	vulnName, vulnLink string
+	uniqueVulnsIdx     map[string]int
+	fixedVulns         []claircore.Vulnerability
+	c                  *csaf.CSAF
+	pc                 *productCache
+	rc                 *repoCache
+}
+
+func (c *creator) knownAffectedVulnerabilities(ctx context.Context, v csaf.Vulnerability, protoVulnFunc func() *claircore.Vulnerability) ([]*claircore.Vulnerability, error) {
+	// Known affected
+	out := []*claircore.Vulnerability{}
+	for _, pc := range v.ProductStatus["known_affected"] {
+		rel := c.c.ProductTree.Relationships.FindRelationship(pc, "default_component_of")
+		if rel == nil {
+			// Should never get here, error in data
+			zlog.Debug(ctx).
+				Str("product:package", pc).
+				Msg("could not find a relationship for product:package")
+			continue
+		}
+		if strings.HasPrefix(rel.ProductRef, "kernel") {
```

Comment:
Ask Prod Sec about this once, and I see we wrote this note:

> For Scanner Certification purposes, it’s ok to show kernel header vulns, but you can also feel free to ignore them since they aren’t used in images

so LGTM